### PR TITLE
[CST-5674] Edit resource policy

### LINKED ISSUE #1585: Edit of resource policy set some fields as readonly without a business reason
**Describe the bug**
In the edit resource policy screen the resource policy type, Action and eperson/Group fields are set in readonly mode.
This was due to lack of support on the rest side but from an user perspective it makes no sense.
Moreover, if you want to edit an INHERITED policy it would be more correct to switch the policy to a TYPE_CUSTOM as this would be the result of the same operation if done at the submission time.
Changing the Action type should be a not very frequent/real scenario but again there is no real obstacle in supporting it.

The group/eperson fields are a bit more challenging as to switch from a resourcepolicy grant to a group to one target to an eperson two path operations would be required and the resourcepolicy will be in an inconsistent state in the middle (no eperson, no group associated). Nevertheless, changing the group associated to a group based resourcepolicy would be not controversial and it is likely to be the most frequent use case (to switch from restricted to administrator to open access - anonymous or similar).

**Expected behavior**
Resource policy type, Action and current eperson or group selection should be editable. A message should inform the user that the policy cannot be moved from a group to an eperson or vicevera and tell him to delete the policy and create a new one if this is the case.
A corresponding REST Ticket should be created to implement the relevant patch methods. For the eperson/group fields only the replace operation should be supported so that there is no way to support operations that could put the resourcepolicy in an inconsistent state.

### PR DESCRIPTION
**Pull request by davide-negretti**
## References
* Fixes #1585
* Requires DSpace/DSpace#8285
* Requires DSpace/DSpace#8287

## Description
It is now possible to change the **policy type**, the **action type** and the **target** (ePerson or group) of a resource policy.

image

## Important warnings

* **Due to a pre-existing bug the new policy target will not be visible in the **Edit bitstream's Policies** page (a page refresh is needed).**  
#644
* When the form is submitted, two separate HTTP requests are done: a PATCH for the policy details and a POST for the new policy target. It may happen that one request fails while the other one doesn't. In case of error, the notification will inform the user about the specific errors that have occurred.

## Instructions for Reviewers

List of changes in this PR:
* Policy type, action type and target fields were previously enabled only when a new policy was being created. They are now enabled also when an existing policy is being edited. 
* An `updateTarget()` method has been implemented to update the policy target

## Checklist
- [x] My PR is small in size (e.g. less than 1,000 lines of code, not including comments & specs/tests), or I have provided reasons as to why that's not possible.
- [x] My PR passes [TSLint](https://palantir.github.io/tslint/) validation using `yarn run lint`
- [x] My PR doesn't introduce circular dependencies
- [x] My PR includes [TypeDoc](https://typedoc.org/) comments for _all new (or modified) public methods and classes_. It also includes TypeDoc for large or complex private methods.
- [x] My PR passes all specs/tests and includes new/updated specs or tests based on the [Code Testing Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Testing+Guide).
- [x] If my PR includes new, third-party dependencies (in `package.json`), I've made sure their licenses align with the [DSpace BSD License](https://github.com/DSpace/DSpace/blob/main/LICENSE) based on the [Licensing of Contributions](https://wiki.lyrasis.org/display/DSPACE/Code+Contribution+Guidelines#CodeContributionGuidelines-LicensingofContributions) documentation.

### CODE REVIEW

**Review by corrad82-4s — CHANGES_REQUESTED** [review]
I have tested this PR, together with its rest counterpart, and all seems fine to me.
I'd just fix the alerts raised by LGTM engine, and check why there's a failing test just with node 12, whereas all tests are passing with node 14.

**Comment by tdonohue** [comment]
@davide-negretti and @corrad82-4s : When a test fails _only on_ Node 12 or _only on_ Node 14, that is **usually** a sign that a random hiccup occurred.  In this case, the Node 12 tests look like they couldn't run because headless Chrome failed to start in our CI environment.  So, I restarted the Node 12 tests a moment ago & I'd expect they probably will succeed this time around.

**Review by corrad82-4s — APPROVED** [review]
Hi,
since we have clarified the issue occurring with node 12 tests, and lgtm reports have been solved. This PR looks good to me

File: src/app/core/resource-policy/resource-policy.service.ts [hunk 1]
```diff
@@ -221,4 +233,43 @@ export class ResourcePolicyService {
...
+    headers = headers.append('Content-Type', 'text/uri-list');
+    options.headers = headers;
+
+    const requestId = this.requestService.generateRequestId();
+
+    this.requestService.setStaleByHrefSubstring(`${this.dataService.getLinkPath()}/${resourcePolicyId}/${type}`);
+
+    return targetEndpoint$.pipe(switchMap((targetEndpoint) => {
+      const resourceEndpoint = resourcePolicyHref + '/' + type;
+      const request = new PostRequest(requestId, resourceEndpoint, targetEndpoint, options);
```
Comment (abollini) [inline src/app/core/resource-policy/resource-policy.service.ts:263]: you need to switch to a PUT request, see comments in the rest contract

**Review by abollini — CHANGES_REQUESTED** [review]
changes are required due to the rest contract feedback, see https://github.com/DSpace/RestContract/pull/193

**Comment by davide-negretti** [comment]
@abollini @corrad82-4s @tdonohue code has been updated according to the changes to the REST contract

File: src/app/core/resource-policy/resource-policy.service.ts [hunk 2]
```diff
@@ -221,4 +233,43 @@ export class ResourcePolicyService {
...
   }
 
+  /**
+   * Update the target of the resource policy
+   * @param resourcePolicyId the ID of the resource policy
+   * @param resourcePolicyHref the link to the resource policy
+   * @param uuid the UUID of the target to which the permission is being granted
+   * @param type the type of the target (eperson or group) to which the permission is being granted
+   */
+  updateTarget(resourcePolicyId: string, resourcePolicyHref: string, uuid: string, type: string): Observable<RemoteData<any>> {
```
Comment (tdonohue) [inline src/app/core/resource-policy/resource-policy.service.ts:243]: I'm not seeing new tests in `resource-policy.service.spec.ts` that correspond to this new method.  Could we add tests for it?

**Review by tdonohue — CHANGES_REQUESTED** [review]
@davide-negretti : Overall the code looks good & it works.  However, I'm seeing a (possible caching) issue immediately after I update any policy. Here's what I'm doing:

1. First, I'm running this in production mode (`yarn build:prod; yarn serve:ssr`)
2. Edit an existing Item. Go to its Authorization policies
3. Find the "Anonymous READ" policy for the Item.  Edit it, and change to "Administrator" group.  Click save.  
4. Update is successful, but when I'm sent back to the list of policies, the READ policy has no Group listed (see screenshot). 
5. If I reload the page (click reload in browser), it updates and shows "Administrator" group correctly.
missing-group

Beyond that strange caching issue, the functionality works.  I also have an inline comment below where it looks like we might be missing new tests.  If it's possible to add those easily, I'd appreciate it.

**Comment by davide-negretti** [comment]
@tdonohue 
> Overall the code looks good & it works. However, I'm seeing a (possible caching) issue immediately after I update any policy
This is a known issue that has already been reported two years ago, see "Important warnings" in the original PR comment for the original GitHub issue.

> Beyond that strange caching issue, the functionality works. I also have an inline comment below where it looks like we might be missing new tests. If it's possible to add those easily, I'd appreciate it.
Ok, I will try to add some coverage

**Comment by davide-negretti** [comment]
@tdonohue I suggest to test this again after having merged #1608, it is possible that this will solve the caching issue

**Comment by davide-negretti** [comment]
@tdonohue I added a new test case for `updateTarget` method

**Review by tdonohue — APPROVED** [review]
👍 Thanks @davide-negretti !  Re-tested / re-reviewed today, and this looks good enough for me.  I can still reproduce the caching issue caused by #644 which I described in [my prior review](https://github.com/DSpace/dspace-angular/pull/1641#pullrequestreview-977195916).  But, that doesn't seem to be the fault of this PR.  So, I'll go ahead and approve this PR & merge it -- and I flagged #644 as `high priority` so that hopefully we can find a solution in 7.4

(Side note: I also verified that #741 does NOT solve this specific caching issue... though that PR is also beneficial, this caching bug is a different issue it seems.)

**Comment by tdonohue** [comment]
Merging with +3

**Pull request merged**